src/pubsub.ts

```diff
@@ -541,32 +546,125 @@ export class PubSub {
       }
     );
   }
+
+  detachSubscription(
+    name: string,
+    gaxOpts?: CallOptions
+  ): Promise<DetachSubscriptionResponse>;
+  detachSubscription(name: string, callback: DetachSubscriptionCallback): void;
+  detachSubscription(
+    name: string,
+    gaxOpts: CallOptions,
```

Comment:
Usually if `gaxOpts` is optional in one invocation model (promises), it can be optional in the callback model as well.  little weird that you MUST specify `gaxOpts` only for callbacks.

Side crazy question - what if we just like didn't support callbacks for new APIs? May be a good way to try this out and measure the response.

Reply:
Yeah, I was again just matching existing style. There was some discussion about dropping callbacks earlier, and I think that was unpopular. I'm up for discussing it though.

Reply:
Hmm, I see what was going on here:
```
  detachSubscription(
    name: string,
    gaxOpts?: CallOptions
  ): Promise<DetachSubscriptionResponse>;
  detachSubscription(name: string, callback: DetachSubscriptionCallback): void;
  detachSubscription(
    name: string,
    gaxOpts: CallOptions,
    callback: DetachSubscriptionCallback
  ): void;
  detachSubscription(
    name: string,
    optsOrCallback?: CallOptions | DetachSubscriptionCallback,
    callback?: DetachSubscriptionCallback
  ): Promise<DetachSubscriptionResponse> | void {
```

Basically this is just TypeScript resolution shenanigans. If you call with only gaxOpts, you want it to give you a Promise. If you call with gaxOpts and a callback, you get the second variant. If you call with only a callback, it will fall through to the last one, which has optsOrCallback. I don't think it has an effect one way or another on the runtime behavior, but it's sort of gross... but I get why Dave did it that way, too.

My "want a pony" goals for the upcoming quarters include switching everything to use Promise internally, and wrap stuff with callbackify(). There are a few things I need to figure out first though. (Like methods that just pass through callbacks directly.)